### ISSUE TITLE
Hikari connection pool does not evict closed connections? (ver.3.4.1)

### ISSUE THREAD

**Issue by kolsan789** [brettwooldridge/HikariCP#1705]
We have setup in AWS with two AZ instances of Aurora PostgreSQL. One of the instances (AZ 1) was switched off recently and the connection pool didn't recover to the second AZ for at least 8 hours. After manual app restart  everything got to normal. 
SpringBoot app with conf
maximum-pool-size: 100
max life time (not set) hence default - 30 mins
It seems to me the connection is not evicted from the pool  although "PoolBase.isConnectionAlive" fails and returns false

I got this error message for this period:
............
Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:697)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:196)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:161)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
	at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)
	at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:38)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:104)
	... 147 common frames omitted
Caused by: org.postgresql.util.PSQLException: This connection has been closed.
	at org.postgresql.jdbc.PgConnection.checkClosed(PgConnection.java:782)
	at org.postgresql.jdbc.PgConnection.setNetworkTimeout(PgConnection.java:1560)
	at com.zaxxer.hikari.pool.PoolBase.setNetworkTimeout(PoolBase.java:552)
	at com.zaxxer.hikari.pool.PoolBase.isConnectionAlive(PoolBase.java:168)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:185)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:161)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
	at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)..........

**Comment by NaldoRay** [comment]
Our service also experiences this similar issue during high db load including during `pg_basebackup`.
Hikari does not recover closed connections/cannot acquire new connections even 12+ hours after the connections has been closed and load is low.
After the load has decreased and there's subsequent db queries, Hikari warns that it's failed to validate connections because the connections has been closed. Subsequent pool stats shows total connection it's become 0. And subsequent queries will failed because connection is not available.

No connection leak detected. All transactions are done under 1s during this time.

These logs show the symptoms. HikariPool give lots of warning "failed to validate" messages.
```
........
[2020-12-19T20:25:24]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@854017b (This connection has been closed.)
[2020-12-19T20:25:24]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@c6b842e (This connection has been closed.)	
[2020-12-19T20:25:24]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@2ac6a0b (This connection has been closed.)
[2020-12-19T20:25:24]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@22f5c280 (This connection has been closed.)
[2020-12-19T20:25:24]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@3889a865 (This connection has been closed.)
[2020-12-19T20:25:21]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@3bc9c4be (This connection has been closed.)
[2020-12-19T20:25:24] DEBUG [service,,,] 8414 --- [HikariPool-1 connection adder] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - After adding stats (total=9, active=1, idle=8, waiting=0)
[2020-12-19T20:25:23] DEBUG [service,,,] 8414 --- [HikariPool-1 connection adder] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Cannot acquire connection from data source
org.postgresql.util.PSQLException: The connection attempt failed.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:292)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:211)
	at org.postgresql.Driver.makeConnection(Driver.java:458)
	at org.postgresql.Driver.connect(Driver.java:260)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:117)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:123)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:375)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:204)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:459)
	at com.zaxxer.hikari.pool.HikariPool.access$200(HikariPool.java:70)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:696)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:682)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.postgresql.core.PGStream.<init>(PGStream.java:75)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:91)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:192)
	... 16 common frames omitted
```

After that, connection adder is still able to add few connections. 
But 1 sec after that HikariPool warns lots of validation failed messages until the pool finally becomes empty:
```
[2020-12-19T20:25:24] DEBUG [service,,,] 8414 --- [HikariPool-1 connection adder] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - After adding stats (total=10, active=2, idle=8, waiting=0)
[2020-12-19T20:25:24] DEBUG [service,,,] 8414 --- [HikariPool-1 connection adder] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@6562cbc9
..................
[2020-12-19T20:25:27] DEBUG [service,,,] 8414 --- [HikariPool-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Pool stats (total=12, active=2, idle=10, waiting=0)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@20021991 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@2ec5e3e1 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@2afa72e1 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@6ff8e9da (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@5f06ab25 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@5e5d81cf (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@27499d9f (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@fd71695 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@6562cbc9 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@57837d53 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-0-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@3c4db759 (This connection has been closed.)
[2020-12-19T20:25:28]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@3f89ab00 (This connection has been closed.)
[2020-12-19T20:25:57] DEBUG [service,,,] 8414 --- [HikariPool-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Pool stats (total=0, active=0, idle=0, waiting=2)
```

Afterwards, subsequent queries will be unable to acquire connection until we restart the service:
```
[2020-12-19T20:26:08] DEBUG [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Timeout failure stats (total=0, active=0, idle=0, waiting=1)
[2020-12-19T20:26:08]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 08003
[2020-12-19T20:26:08] ERROR [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] o.h.engine.jdbc.spi.SqlExceptionHelper   : HikariPool-1 - Connection is not available, request timed out after 60001ms.
[2020-12-19T20:26:08]  WARN [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 08003
[2020-12-19T20:26:08] ERROR [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] o.h.engine.jdbc.spi.SqlExceptionHelper   : This connection has been closed.
[2020-12-19T20:26:08] ERROR [service,,,] 8414 --- [org.springframework.kafka.KafkaListenerEndpointContainer#9-1-L-17] o.s.kafka.listener.LoggingErrorHandler   : Error while processing: ConsumerRecord ......
Caused by: org.springframework.dao.DataAccessResourceFailureException: Unable to acquire JDBC Connection; nested exception is org.hibernate.exception.JDBCConnectionException: Unable to acquire JDBC Connection
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:257)
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244)
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:488)
	at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:59)
	at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:213)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:147)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:133)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:57)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
	at com.sun.proxy.$Proxy204.findByAirwayBillNoAndPackageId(Unknown Source)
	at com.xyz.AbcListener.onAbcEvent(AbcListener.java:50)
	at sun.reflect.GeneratedMethodAccessor262.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:180)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:112)
	at org.springframework.kafka.listener.adapter.HandlerAdapter.invoke(HandlerAdapter.java:48)
	at org.springframework.kafka.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:174)
	... 9 common frames omitted
Caused by: org.hibernate.exception.JDBCConnectionException: Unable to acquire JDBC Connection
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:48)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:97)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:109)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.getPhysicalConnection(LogicalConnectionManagedImpl.java:136)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.connection(StatementPreparerImpl.java:47)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:146)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:172)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:148)
	at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:1985)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1915)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1893)
	at org.hibernate.loader.Loader.doQuery(Loader.java:938)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:341)
	at org.hibernate.loader.Loader.doList(Loader.java:2692)
	at org.hibernate.loader.Loader.doList(Loader.java:2675)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2507)
	at org.hibernate.loader.Loader.list(Loader.java:2502)
	at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:502)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:392)
	at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:216)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1490)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1445)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1414)
	at org.hibernate.query.internal.AbstractProducedQuery.getSingleResult(AbstractProducedQuery.java:1463)
	at sun.reflect.GeneratedMethodAccessor264.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.orm.jpa.SharedEntityManagerCreator$DeferredQueryInvocationHandler.invoke(SharedEntityManagerCreator.java:372)
	at com.sun.proxy.$Proxy291.getSingleResult(Unknown Source)
	at org.springframework.data.jpa.repository.query.JpaQueryExecution$SingleEntityExecution.doExecute(JpaQueryExecution.java:208)
	at org.springframework.data.jpa.repository.query.JpaQueryExecution.execute(JpaQueryExecution.java:87)
	at org.springframework.data.jpa.repository.query.AbstractJpaQuery.doExecute(AbstractJpaQuery.java:116)
	at org.springframework.data.jpa.repository.query.AbstractJpaQuery.execute(AbstractJpaQuery.java:106)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:483)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:461)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136)
	... 26 common frames omitted
Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 60001ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:666)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:182)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:147)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
	at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)
	at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:35)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:106)
	... 66 common frames omitted
Caused by: org.postgresql.util.PSQLException: This connection has been closed.
	at org.postgresql.jdbc.PgConnection.checkClosed(PgConnection.java:857)
	at org.postgresql.jdbc.PgConnection.setNetworkTimeout(PgConnection.java:1639)
	at com.zaxxer.hikari.pool.PoolBase.setNetworkTimeout(PoolBase.java:551)
	at com.zaxxer.hikari.pool.PoolBase.isConnectionAlive(PoolBase.java:172)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:171)
	... 71 common frames omitted
```

TL;DR
Database is under high load, 80-100% cpu usage.
Connections are closed from DB side. 
Hikari failed to validate and acquire connections even 12+ hours after connections have been closed and load is low.
Connection pool back to normal after restarting the service.

**Comment by nunnrlc** [comment]
Hi,

To answer the title...
> Hikari connection pool does not evict closed connections

It seems the connections *are* evicted, but HikariCP can't open new ones. The issue description doesn't match the title I think?

RE: changing AZ of the database...
You haven't detailed, but I assume Spring is configured to connect to a named domain? Then the DNS record is updated when the DB server is changed?
Thing is, DNS resolution is may be cached by Java until shutdown, so the IP address change will not be picked up. That would cause the issue described - not connecting to the new server until restart.

In this case please configure DNS TTL for Java.

See AWS docs here for full details: https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-jvm-ttl.html

From inside your app it is one line to add before your application start up:
`java.security.Security.setProperty("networkaddress.cache.ttl" , "60");`

If you are sure the DNS change is picked up by Java, then please add more details about how your app is configured.

Poster # 2:
pg_basebackup shouldn't be closing existing client connections, so that's strange. Are you sure no other server maintenance or rotation is performed when the backup is run? Again, if the server is switched over, HikariCP will continue to try the old server...

**Comment by kolsan789** [comment]
Hi,

Our configuration regarding networkaddress.cache.ttl is 30 secs. 
The url is configured like this spring.datasource.url= jdbc:postgresql://rds.cluster.name:5432/databsename (no other connection options set)
Hikari pool size spring.datasource.hikari.maximum-pool-size= 100

The interesting thing is that we cannot reproduce the issue when try manual switching of the instances. In my tests the connections get recovered in about 1 min. So the issue has happened only once and haven't seen it afterwards

**Comment by philippn** [comment]
Hello,

I have observed the same problem on a IBM cloud postgres database with Hikari 3.45. According to the metrics, the available connections and idle connections effectively plummeted to 0 and the log gets spammed with error messages. Load was low at the time when it started. This was also with an AdoptOpenJDK 11 on Alpine, so I believe the AdoptOpenJDK default of ttl=30 seconds should be in effect here.

Could it be that some other component other than the JVM itself is also caching DNS lookups? I.e. the Postgres driver?

Any help would be appreciated

**Comment by KazzmanK** [comment]
Just experienced same issue with MSSQL while VM with DB hang and hard restarted. 
Log is full with Connection closed exceptions, and no new connection was made.

com.zaxxer:HikariCP:jar:5.0.1    used

```
Caused by: java.sql.SQLTransientConnectionException: hikariCPPool - Connection is not available, request timed out after 30001ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:181)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
	at org.springframework.jdbc.datasource.DataSourceUtils.fetchConnection(DataSourceUtils.java:160)
	at org.springframework.jdbc.datasource.DataSourceUtils.doGetConnection(DataSourceUtils.java:118)
	at org.springframework.jdbc.datasource.DataSourceUtils.getConnection(DataSourceUtils.java:81)
	... 105 common frames omitted
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Connection closed.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDriverError(SQLServerException.java:237)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.checkClosed(SQLServerConnection.java:1555)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.setNetworkTimeout(SQLServerConnection.java:6648)
	at com.zaxxer.hikari.pool.PoolBase.setNetworkTimeout(PoolBase.java:561)
	at com.zaxxer.hikari.pool.PoolBase.isConnectionDead(PoolBase.java:168)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:170)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
```

**Comment by goelvipul27** [comment]
Hi All,

Facing the same issue with Postgres Hyperscalars.
I have tried "tcpKeepAlive" as true.
using 
hikariConfig.addDataSourceProperty("tcpKeepAlive",true);
This resolved the issue.

Please let me know your thoughts on the same.

tcpKeepAlive is a PostgreSQL driver-specific property; one has to find a similar property for their respective driver.

TCP Keepalive packets ensure that idle TCP connections remain open by periodically sending data between the client and server. If a response to a keepalive packet is not received, the connection is terminated 
How it works:- TCP Keepalive packets ensure that idle TCP connections remain open by periodically sending data between the client and server. If a response to a keepalive packet is not received, the connection is terminated.

When enabled, the JDBC driver periodically sends keepalive packets to the server. The server responds to these packets with acknowledgements. If the server doesn't respond to a keepalive packet, the connection is considered inactive and is closed.

Thanks,
Vipul
